Typecheck fails because of `virtual:react-router/server-build`

### I'm using React Router as a...

framework

### Reproduction

[stackblitz](https://stackblitz.com/edit/vitejs-vite-lp6hibgv?file=package.json)

### System Info

```shell
System:
    OS: macOS 15.3.2
    CPU: (8) arm64 Apple M1 Pro
    Memory: 85.94 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
Binaries:
    Node: 20.12.2 - ~/.nvm/versions/node/v20.12.2/bin/node
    npm: 10.9.0 - ~/.nvm/versions/node/v20.12.2/bin/npm
    pnpm: 10.5.0 - ~/Library/pnpm/pnpm
    Watchman: 2024.08.26.00 - /opt/homebrew/bin/watchman
Browsers:
    Chrome: 134.0.6998.89
    Edge: 134.0.3124.68
    Safari: 18.3.1
npmPackages:
    @react-router/dev: ^7.4.0 => 7.4.0 
    @react-router/node: ^7.4.0 => 7.4.0 
    react-router: ^7.4.0 => 7.4.0 
    vite: ^6.2.2 => 6.2.2
```

### Used Package Manager

pnpm

### Expected Behavior

typecheck should not emit errors tied to `virtual:react-router/server-build` and `react-router` in `+virtual.ts` file

it passes when `exactOptionalPropertyTypes` is enabled like it's mentioned [here](https://github.com/remix-run/react-router/pull/13152#issuecomment-2737446847) but then a lot of different errors pop up throughout the codebase because of `exactOptionalPropertyTypes`

### Actual Behavior

typecheck on v7.4.0 produces these two errors: 

```
.react-router/types/+virtual.ts:1:16 - error TS2664: Invalid module name in augmentation, module 'virtual:react-router/server-build' cannot be found.

declare module "virtual:react-router/server-build" {

.react-router/types/+virtual.ts:2:31 - error TS2307: Cannot find module 'react-router' or its corresponding type declarations.

import { ServerBuild } from "react-router";
```

--- a/.changeset/honest-moles-brush.md
+++ b/.changeset/honest-moles-brush.md
@@ -1,0 +1,5 @@
+---
+"@react-router/dev": patch
+---
+
+Fix typegen for virtual modules when `moduleDetection` is set to `force`

--- a/integration/typegen-test.ts
+++ b/integration/typegen-test.ts
@@ -466,6 +466,30 @@ test.describe("typegen", () => {
       expect(proc.status).toBe(0);
     });
 
+    test("works with tsconfig 'moduleDetection' set to 'force'", async () => {
+      const cwd = await createProject({
+        "vite.config.ts": viteConfig,
+        "app/routes.ts": tsx`
+          import { type RouteConfig } from "@react-router/dev/routes";
+          export default [] satisfies RouteConfig;
+        `,
+        "app/handler.ts": tsx`
+          import { createRequestHandler } from "react-router";
+          import * as serverBuild from "virtual:react-router/server-build";
+          export default createRequestHandler(serverBuild);
+        `,
+      });
+
+      const tsconfig = await fse.readJson(path.join(cwd, "tsconfig.json"));
+      tsconfig.compilerOptions.moduleDetection = "force";
+      await fse.writeJson(path.join(cwd, "tsconfig.json"), tsconfig);
+
+      const proc = typecheck(cwd);
+      expect(proc.stdout.toString()).toBe("");
+      expect(proc.stderr.toString()).toBe("");
+      expect(proc.status).toBe(0);
+    });
+
     test("dynamic import matches 'createRequestHandler' function argument type", async () => {
       const cwd = await createProject({
         "vite.config.ts": viteConfig,

--- a/packages/react-router-dev/typegen/index.ts
+++ b/packages/react-router-dev/typegen/index.ts
@@ -89,7 +89,7 @@ async function writeAll(ctx: Context): Promise<void> {
   const registerPath = Path.join(typegenDir, "+register.ts");
   fs.writeFileSync(registerPath, register(ctx));
 
-  const virtualPath = Path.join(typegenDir, "+virtual.ts");
+  const virtualPath = Path.join(typegenDir, "+virtual.d.ts");
   fs.writeFileSync(virtualPath, virtual);
 }
 

